sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -389,7 +389,7 @@ case class InMemoryRelation(
 
   @volatile var statsOfPlanToCache: Statistics = null
 
-  override def innerChildren: Seq[SparkPlan] = Seq(cachedPlan)
+  override val innerChildren: Seq[SparkPlan] = Seq(cachedPlan.clone())
```

Comment:
Does this mean the cloned `cachedPlan` can still be shared by multiple threads?

Reply:
Since plans are considered immutable and usually cloned, thread safety is not an issue. `cachedPlan` was not cloned so it was easy to get in a situation where multiple threads access the same plan objects. This solves that narrow issue. After this the `cachedPlan` is no more susceptible to concurrency issues than any other object in the plan. 

I still think there is a wider thread safety issue with the mutable state in `TreeNode.tags`, but I need to investigate it more to propose a holistic solution. Consider the case where in the application code:

1. User thread 1 calls `d.explain("formatted")` where `d` is a `Dataset` object. This will set the `TreeNode.tag` keys `QueryPlan.OP_ID_TAG` and `QueryPlan.CODEGEN_ID_TAG`. 
2. At the same time in user thread 2 they call `d.filter($"foo" > 1)`. This will create a new dataset that reads the `TreeNode.tags` during the clone. 

At first I considered making `TreeNode.tags` a thread-safe map, but that only solves the concurrency issue. At least for this issue, there would still be a correctness issue with other threads setting different operator ids for the same node. `TreeNode.tags` is used in so many other places that I'm not sure if there are other correctness issues.

Reply:
@cloud-fan friendly ping on this